File: src/faebryk/core/zig/test/sexp/performance/performance_sexp.zig
const std = @import("std");
const sexp = @import("sexp");

const TokenCounts = struct {
    lparen: usize = 0,
    rparen: usize = 0,
    symbols: usize = 0,
    numbers: usize = 0,
    strings: usize = 0,
    comments: usize = 0,
};

fn countTokenTypes(tokens: []const sexp.tokenizer.Token) TokenCounts {
    var counts = TokenCounts{};

    for (tokens) |token| {
        switch (token.type) {
            .lparen => counts.lparen += 1,
            .rparen => counts.rparen += 1,
            .symbol => counts.symbols += 1,
            .number => counts.numbers += 1,
            .string => counts.strings += 1,
            .comment => counts.comments += 1,
        }
    }

    return counts;
}

const SExpCounts = struct {
    lists: usize = 0,
    atoms: usize = 0,
};

fn countSExpElements(sexp_ast: ?sexp.ast.SExp) SExpCounts {
    var counts = SExpCounts{};

    if (sexp_ast) |s| {
        countSExpRecursive(s, &counts);
    }

    return counts;
}

fn countSExpRecursive(sexp_ast: sexp.ast.SExp, counts: *SExpCounts) void {
    switch (sexp_ast.value) {
        .list => |items| {
            counts.lists += 1;
            for (items) |item| {
                countSExpRecursive(item, counts);
            }
        },
        .symbol, .number, .string, .comment => counts.atoms += 1,
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    const arena_allocator = arena.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 2) {
        std.debug.print("Usage: {s} <file.kicad_pcb>\n", .{args[0]});
        std.process.exit(1);
    }

    const file_path = args[1];

    // Get CPU count
    const cpu_count = try std.Thread.getCpuCount();
    std.debug.print("CPU count: {}\n", .{cpu_count});

    var timer = try std.time.Timer.start();

    // Read file size
    const file = try std.fs.cwd().openFile(file_path, .{});
    const file_size = try file.getEndPos();
    file.close();

    std.debug.print("File size: {} bytes ({d:.2} MB)\n", .{ file_size, @as(f64, @floatFromInt(file_size)) / (1024.0 * 1024.0) });
    std.debug.print("\n", .{});

    // Test1: Read file content
    const file_content = try std.fs.cwd().readFileAlloc(allocator, file_path, 100 * 1024 * 1024);
    defer allocator.free(file_content);
    const read_time = timer.read();
    std.debug.print("Read time: {d:.3} ms\n", .{@as(f64, @floatFromInt(read_time)) / 1_000_000.0});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(read_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});

    // Test2: File tokenization
    std.debug.print("=== Tokenization ===\n", .{});
    timer.reset();

    const tokens = try sexp.tokenizer.tokenize(allocator, file_content);
    defer allocator.free(tokens);

    const token_time = timer.read();
    std.debug.print("Time: {d:.3} ms\n", .{@as(f64, @floatFromInt(token_time)) / 1_000_000.0});
    std.debug.print("Tokens: {}\n", .{tokens.len});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(token_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});
    std.debug.print("\n", .{});

    // Test 3: Full parsing with arena allocator
    std.debug.print("=== Full Parsing (Arena Allocator) ===\n", .{});

    timer.reset();
    const sexp_ast = try sexp.ast.parse(arena_allocator, tokens);
    const parse_time = timer.read();

    timer.reset();
    var structure_time: u64 = 0;
    // Only perform structure decoding if the file is a .net file
    if (std.mem.endsWith(u8, file_path, ".net")) {
        var netlistfile = try sexp.kicad.netlist.NetlistFile.loads(arena_allocator, .{ .sexp = sexp_ast });
        defer netlistfile.free(arena_allocator);
        structure_time = timer.read();
    } else if (std.mem.endsWith(u8, file_path, ".kicad_pcb")) {
        var pcb_file = try sexp.kicad.pcb.PcbFile.loads(arena_allocator, .{ .sexp = sexp_ast });
        defer pcb_file.free(arena_allocator);
        structure_time = timer.read();
        std.debug.print("{}\n", .{pcb_file.kicad_pcb.footprints.len});
    }

    std.debug.print("Tokenize time: {d:.3} ms\n", .{@as(f64, @floatFromInt(token_time)) / 1_000_000.0});
    std.debug.print("Parse time: {d:.3} ms\n", .{@as(f64, @floatFromInt(parse_time)) / 1_000_000.0});
    std.debug.print("Structure time: {d:.3} ms\n", .{@as(f64, @floatFromInt(structure_time)) / 1_000_000.0});
    std.debug.print("Total time (tokenize + parse + structure): {d:.3} ms\n", .{@as(f64, @floatFromInt(token_time + parse_time + structure_time)) / 1_000_000.0});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(token_time + parse_time + structure_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});

    // Output in format expected by Python script
    const counts = countTokenTypes(tokens);
    const sexp_counts = countSExpElements(sexp_ast);
    const tokenize_time_ms = @divFloor(token_time, 1_000_000);
    const parse_time_ms = @divFloor(parse_time, 1_000_000);
    const structure_time_ms = @divFloor(structure_time, 1_000_000);

    // RESULT:file_size:tokenize_time:parse_time:total_tokens:lparen:rparen:symbols:numbers:strings:comments:balanced:sexp_count:list_count:atom_count:cpu_count
    std.debug.print("RESULT:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}\n", .{
        file_size,
        tokenize_time_ms,
        parse_time_ms,
        structure_time_ms,
        tokens.len,
        counts.lparen,
        counts.rparen,
        counts.symbols,
        counts.numbers,
        counts.strings,
        counts.comments,
        counts.lparen == counts.rparen,
        //if (sexp != null) @as(usize, 1) else @as(usize, 0),
        @as(usize, 1),
        sexp_counts.lists,
        sexp_counts.atoms,
        cpu_count,
    });
}

File: src/faebryk/core/zig/test/sexp/performance/performance_sexp.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.8"
# dependencies = [
#     "sexpdata",
#     "atopile"
# ]
# ///

import subprocess
import time
from dataclasses import dataclass
from pathlib import Path

import sexpdata

from faebryk.libs.kicad.fileformats import kicad

SEXP_ROOT = Path(__file__).parent.parent.parent
TEST_FILES_DIR = SEXP_ROOT / "test" / "resources" / "v9"


def count_tokens(data, depth=0):
    """Recursively count tokens and S-expressions in parsed data"""
    lparen = 0
    rparen = 0
    symbols = 0
    numbers = 0
    strings = 0
    lists = 0
    atoms = 0

    def count_recursive(item):
        nonlocal lparen, rparen, symbols, numbers, strings, lists, atoms

        if isinstance(item, list):
            lparen += 1
            rparen += 1
            lists += 1
            for sub_item in item:
                count_recursive(sub_item)
        elif isinstance(item, sexpdata.Symbol):
            symbols += 1
            atoms += 1
        elif isinstance(item, (int, float)):
            numbers += 1
            atoms += 1
        elif isinstance(item, str):
            strings += 1
            atoms += 1

    count_recursive(data)
    return (
        lparen,
        rparen,
        symbols,
        numbers,
        strings,
        0,
        lists,
        atoms,
    )  # 0 comments - Python parser strips them


def test_python(file_path):
    """Test Python sexpdata performance"""
    # Read file
    file_contents = file_path.read_text()
    file_size = len(file_contents)

    # Parse with timing
    start_time = time.time() * 1000  # milliseconds
    try:
        data = sexpdata.loads(file_contents)
        end_time = time.time() * 1000
        parse_time = int(end_time - start_time)

        if file_path.suffix == ".net":
            _ = kicad.loads(kicad.netlist.NetlistFile, data)
            structure_time = int((time.time() * 1000) - end_time)
        elif file_path.suffix == ".kicad_pcb":
            _ = kicad.loads(kicad.pcb.PcbFile, data)
            structure_time = int((time.time() * 1000) - end_time)
        else:
            structure_time = 0

        # Count tokens
        lparen, rparen, symbols, numbers, strings, comments, lists, atoms = (
            count_tokens(data)
        )
        total_tokens = lparen + rparen + symbols + numbers + strings + comments

        return TestResults(
            file_size=file_size,
            tokenize_time=0,  # Python does both in one pass
            parse_time=parse_time,
            structure_time=structure_time,
            total_tokens=total_tokens,
            lparen=lparen,
            rparen=rparen,
            symbols=symbols,
            numbers=numbers,
            strings=strings,
            comments=comments,
            balanced=lparen == rparen,
            sexp_count=1 if isinstance(data, list) else 0,
            list_count=lists,
            atom_count=atoms,
            cpu_count=1,
        )
    except Exception as e:
        raise Exception(f"Python test failed: {e}")


@dataclass
class TestResults:
    file_size: int
    tokenize_time: int
    parse_time: int
    structure_time: int
    total_tokens: int
    lparen: int
    rparen: int
    symbols: int
    numbers: int
    strings: int
    comments: int
    balanced: bool
    sexp_count: int
    list_count: int
    atom_count: int
    cpu_count: int


def test_zig(file_path):
    """Test Zig parallel tokenizer and AST parser performance"""

    # Run the performance test
    result = subprocess.run(
        ["zig", "build", "perf", "--", str(file_path.absolute())],
        capture_output=True,
        text=True,
        cwd=SEXP_ROOT,
    )

    if result.returncode != 0:
        raise Exception(f"Zig test failed: {result.stderr}")

    # Parse output
    for line in result.stderr.split("\n"):
        if line.startswith("RESULT:"):
            parts = line.replace("RESULT:", "").split(":")
            if len(parts) >= 15:
                return TestResults(
                    file_size=int(parts[0]),
                    tokenize_time=int(parts[1]),
                    parse_time=int(parts[2]),
                    structure_time=int(parts[3]),
                    total_tokens=int(parts[4]),
                    lparen=int(parts[5]),
                    rparen=int(parts[6]),
                    symbols=int(parts[7]),
                    numbers=int(parts[8]),
                    strings=int(parts[9]),
                    comments=int(parts[10]),
                    balanced=parts[11] == "true",
                    sexp_count=int(parts[12]),
                    list_count=int(parts[13]),
                    atom_count=int(parts[14]),
                    cpu_count=int(parts[15]),
                )

    raise Exception("Could not parse Zig output")


def format_results(name, results: TestResults | None):
    if results is None:
        return f"{name}: ERROR - No results"

    """Format test results for display"""
    tokens_str = (
        f"(L:{results.lparen} R:{results.rparen} "
        f"S:{results.symbols} N:{results.numbers} "
        f"Str:{results.strings}"
    )

    # Add comments if present
    if results.comments > 0:
        tokens_str += f" C:{results.comments}"

    tokens_str += ")"

    # Build timing string
    time_str = (
        f"tokenize:{results.tokenize_time}ms "
        f"+ parse:{results.parse_time}ms "
        f"+ structure:{results.structure_time}ms "
        f"= {results.tokenize_time + results.parse_time + results.structure_time}ms"
    )

    # Add S-expression counts if available
    sexp_str = ""
    if results.list_count > 0 and results.atom_count > 0:
        sexp_str = (
            f", S-exps: {results.sexp_count}"
            f" (Lists:{results.list_count} Atoms:{results.atom_count})"
        )

    return (
        f"{name}: {results.file_size} bytes, "
        f"{time_str}, "
        f"{results.total_tokens} tokens "
        # f"{tokens_str} "
        f"{'✓' if results.balanced else '✗'}"
        f"{sexp_str}"
    )


def main():
    # Find all .kicad_pcb files in test_files directory
    test_files = sorted(TEST_FILES_DIR.glob("pcb/*.kicad_pcb"))

    if not test_files:
        print(f"No .kicad_pcb files found in {TEST_FILES_DIR}")
        return

    print(f"Found {len(test_files)} test files\n")

    for test_file in test_files:
        print(f"=== {test_file.name} ===")

        # Test Python
        try:
            python_results = test_python(test_file)
        except Exception:
            python_results = None

        print(format_results("Python", python_results))

        # Test Zig
        try:
            zig_results = test_zig(test_file)
        except Exception:
            zig_results = None

        zig_name = "Zig"
        if zig_results:
            zig_name += f" ({zig_results.cpu_count} cores)"
        print(format_results(zig_name, zig_results))

        # Show speedup if both succeeded
        if python_results and zig_results:
            python_time = (
                python_results.tokenize_time
                + python_results.parse_time
                + python_results.structure_time
            )
            zig_time = (
                zig_results.tokenize_time
                + zig_results.parse_time
                + zig_results.structure_time
            )

            if zig_time > 0:
                speedup = python_time / zig_time
                print(f"Speedup: {speedup:.1f}x")
            else:
                print("Speedup: Too fast to measure accurately")

        # Show throughput
        if zig_results:
            file_size_mb = zig_results.file_size / (1024 * 1024)
            total_time_sec = (
                zig_results.tokenize_time
                + zig_results.parse_time
                + zig_results.structure_time
            ) / 1000.0
            if total_time_sec > 0:
                throughput = file_size_mb / total_time_sec
                print(f"Zig throughput: {throughput:.1f} MB/s")

        print()


if __name__ == "__main__":
    main()
